src/main/java/net/neoforged/neoforge/server/command/TPSCommand.java

```diff
@@ -5,53 +5,53 @@
+import net.minecraft.util.Mth;
 import net.minecraft.util.TimeUtil;
-import net.minecraft.world.level.dimension.DimensionType;
+import net.minecraft.world.TickRateManager;
+import org.jetbrains.annotations.Nullable;
+import org.joml.Math;
 
 class TPSCommand {
     private static final DecimalFormat TIME_FORMATTER = new DecimalFormat("########0.000");
     private static final long[] UNLOADED = new long[] { 0 };
 
     static ArgumentBuilder<CommandSourceStack, ?> register() {
         return Commands.literal("tps")
-                .requires(cs -> cs.hasPermission(0)) //permission
+                .requires(cs -> cs.hasPermission(Commands.LEVEL_ALL)) //permission
                 .then(Commands.argument("dim", DimensionArgument.dimension())
-                        .executes(ctx -> sendTime(ctx.getSource(), DimensionArgument.getDimension(ctx, "dim"))))
+                        .executes(ctx -> sendTime(ctx, DimensionArgument.getDimension(ctx, "dim"))))
                 .executes(ctx -> {
-                    for (ServerLevel dim : ctx.getSource().getServer().getAllLevels())
-                        sendTime(ctx.getSource(), dim);
+                    for (ServerLevel dim : ctx.getSource().getServer().getAllLevels()) {
+                        sendTime(ctx, dim);
+                    }
 
-                    long[] times = ctx.getSource().getServer().getTickTimesNanos();
-                    double meanTickTime = mean(times) * 1.0E-6D;
-                    double meanTPS = TimeUtil.MILLISECONDS_PER_SECOND / Math.max(meanTickTime, ctx.getSource().getServer().tickRateManager().millisecondsPerTick());
-                    ctx.getSource().sendSuccess(() -> Component.translatable("commands.neoforge.tps.summary.all", TIME_FORMATTER.format(meanTickTime), TIME_FORMATTER.format(meanTPS)), false);
-
-                    return 0;
+                    sendTime(ctx, null);
+                    return Command.SINGLE_SUCCESS;
                 });
     }
 
-    private static int sendTime(CommandSourceStack cs, ServerLevel dim) throws CommandSyntaxException {
-        long[] times = cs.getServer().getTickTime(dim.dimension());
-
-        if (times == null) // Null means the world is unloaded. Not invalid. That's taken care of by DimensionArgument itself.
-            times = UNLOADED;
-
-        final Registry<DimensionType> reg = cs.registryAccess().registryOrThrow(Registries.DIMENSION_TYPE);
-        double worldTickTime = mean(times) * 1.0E-6D;
-        double worldTPS = TimeUtil.MILLISECONDS_PER_SECOND / Math.max(worldTickTime, dim.tickRateManager().millisecondsPerTick());
-        cs.sendSuccess(() -> Component.translatable("commands.neoforge.tps.summary.named", dim.getDescription(), reg.getKey(dim.dimensionType()).toString(), TIME_FORMATTER.format(worldTickTime), TIME_FORMATTER.format(worldTPS)), false);
-
-        return 1;
+    private static int sendTime(CommandContext<CommandSourceStack> context, ServerLevel dim) throws CommandSyntaxException {
```

Comment:
The `dim` parameter should also be marked as `@Nullable`, since it's given `null` in the command (for the overall summary).